### PR TITLE
made live on connectcgrduckdns.org/stock-analyzer

### PR DESCRIPTION
**Pull request by cgraaaj**

### CODE REVIEW

File: client/.dockerignore [hunk 1]
```diff
@@ -0,0 +1 @@
+# .env
```
Comment (varadhodiyil) [inline client/.dockerignore:1]: EOF

File: client/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+COPY package.json ./
+RUN npm install
+COPY ./ ./
+RUN npm run build
+
+FROM nginx
+EXPOSE 3000
+COPY ./nginx/default.conf /etc/nginx/conf.d/default.conf
+COPY --from=0 /app/build /usr/share/nginx/html
+CMD ["nginx", "-g", "daemon off;"]
```
Comment (varadhodiyil) [inline client/Dockerfile:12]: EOF

File: server/analyze/getOptionTrend.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def analyze_options_data(self,index, symbol):
+        url = self.option_chain_url.format(index, symbol)
+        try:
+            resp= get_nse_response(url)
+            resp = analyze_stock(resp['records']['expiryDates'][0],resp['records'])
+            # pprint(resp)
+            temp={}
+            temp['name']= symbol
+            temp['options']={"calls":{"bullish":0,"bearish":0},"puts":{"bullish":0,"bearish":0}}
```
Comment (varadhodiyil) [inline server/analyze/getOptionTrend.py:26]: Df.count()

**Review by varadhodiyil — APPROVED** [review]

File: server/analyze/analyze.py [hunk 4]
```diff
@@ -64,10 +70,40 @@ def get_uptrend():
...
+    return "26000"
 
 @analyze.route("/options", methods=["GET"])
 def get_options():
     current_app.logger.info(f"analyzing data")
-    optionTrend = OptionTrend()
-    res = optionTrend.get_option_trend("equities")
-    return jsonify(list(res))
+    mode = "equities"
+    tickers = ['NIFTY','BANKNIFTY','FINNIFTY'] if mode == "indices" else get_nse_response(nse.equities_url)
```
Comment (varadhodiyil) [inline server/analyze/analyze.py:82]: Always false

File: server/analyze/analyze.py [hunk 5]
```diff
@@ -64,10 +70,40 @@ def get_uptrend():
...
+    def generate():
+        for ticker in tickers:
+            yield json.dumps(analyze_options_data(mode, ticker))+'\n'
+    response = current_app.response_class(generate(),mimetype='application/json')
```
Comment (varadhodiyil) [inline server/analyze/analyze.py:86]:
```suggestion
return  current_app.response_class(generate(),mimetype='application/json')
````

File: server/analyze/analyze.py [hunk 6]
```diff
@@ -64,10 +70,40 @@ def get_uptrend():
...
 def get_options():
     current_app.logger.info(f"analyzing data")
-    optionTrend = OptionTrend()
-    res = optionTrend.get_option_trend("equities")
-    return jsonify(list(res))
+    mode = "equities"
+    tickers = ['NIFTY','BANKNIFTY','FINNIFTY'] if mode == "indices" else get_nse_response(nse.equities_url)
+    def generate():
+        for ticker in tickers:
+            yield json.dumps(analyze_options_data(mode, ticker))+'\n'
```
Comment (varadhodiyil) [inline server/analyze/analyze.py:85]: This can possibly yield `None`

File: client/src/components/Uptrend/OptionUptrend.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            setTimeout(() => {
+                document.getElementById("progressBarParent").removeChild(document.getElementById("progressBar"));
+            }, 3000)
+        }
+    }
+
+    handleWindowBeforeUnload = (ev) => {
+        ev.preventDefault();
+        const confirmationMessage = "Do you want to cancel the request";
+        ev.returnValue = confirmationMessage;
```
Comment (varadhodiyil) [inline client/src/components/Uptrend/OptionUptrend.js:41]:
```suggestion
        ev.returnValue = "Do you want to cancel the request";
        
```

File: client/src/components/Uptrend/OptionUptrend.js [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        ));
+    };
+
+    populateGridRow = (options, header) => {
+        options = options.filter((option) =>
+            header === "Call" ? option.callTrend : option.putTrend
+        );
+        options =
+            header === "Call"
+                ? options.sort((a, b) => (a.callTrendDiff > b.callTrendDiff ? -1 : 1))
```
Comment (varadhodiyil) [inline client/src/components/Uptrend/OptionUptrend.js:77]: What should happen when `a.callTrendDiff == b.callTrendDiff` ?

File: server/analyze/analyze.py (hunk 4 above)
Comment (cgraaaj) [reply server/analyze/analyze.py:82]: yes as of now, changes will be made for indices

**Review by varadhodiyil — APPROVED** [review]

File: client/src/actions/index.js [hunk 9]
```diff
@@ -141,16 +141,16 @@ export const setFormValues = (formValues) => {
...
-}
+};
 
 export const getUptrend = () => async (dispatch) => {
   let response = "";
   try {
-    response = await API.get(`/api/analyze/uptrend`)
-    response = response.data
-    console.log(response)
+    response = await API.get(`/api/analyze/uptrend`);
```
Comment (varadhodiyil) [inline client/src/actions/index.js:149]: Endpoints can go in the constants file Aswell

File: client/src/actions/index.js [hunk 10]
```diff
@@ -141,16 +141,16 @@ export const setFormValues = (formValues) => {
...
+    response = response.data;
+    console.log(response);
   } catch (err) {
-    console.log(err)
+    console.log(err);
```
Comment (varadhodiyil) [inline client/src/actions/index.js:153]: Does Success & Failure response have the same payload type?

File: client/src/components/OptionChain/OptionChainTable.js [hunk 11]
```diff
@@ -1,59 +1,72 @@
...
         </div>
       </div>
-    )
+    );
   }
 }
 const mapStateToProps = (state) => {
-  let tableData = JSON.parse(sessionStorage.getItem('tableData'))
-  console.log(tableData)
+  let tableData = JSON.parse(sessionStorage.getItem("tableData"));
```
Comment (varadhodiyil) [inline client/src/components/OptionChain/OptionChainTable.js:62]: Why not [redux](https://stackoverflow.com/questions/56923481/why-redux-instead-of-session-storage)?

File: client/src/components/Uptrend/EquityUptrend.js [hunk 12]
```diff
@@ -1,142 +1,177 @@
...
+        _.isUndefined(this.props.uptrendWithVolume) ? (
+          <div className="ui container">
+            <div
+              style={{ height: "250px", width: "100%" }}
+              className="ui segment"
+            >
+              <div className="ui active inverted dimmer">
+                <div className="ui text loader">Loading</div>
+              </div>
+              <p></p>
```
Comment (varadhodiyil) [inline client/src/components/Uptrend/EquityUptrend.js:112]: ?

**Pull request closed without merging**